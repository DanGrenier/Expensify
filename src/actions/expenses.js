import uuid from 'uuid';
import database from '../firebase/firebase'

//Components call action generator
//action generator returns an object
//component dispatches the object to the reducer
//The reducer changes the store

//ADD EXPENSE
export const addExpense = (expense) => ({
    
    type: 'ADD_EXPENSE',
    expense

    
  });

  export const startAddExpense = (expenseData = {}) => {
      return (dispatch) => {
          const {
              description = '',
              note = '',
              amount = 0,
              createdAt = 0
          } = expenseData;
          const expense = {description,note,amount,createdAt};
          return database.ref('expenses').push(expense).then((ref) => {
              dispatch(addExpense({
                  id: ref.key,
                  ...expense
              }));
          })

      };
  }
  
//REMOVE EXPENSE  
export const removeExpense = ({id} = {}) => ( {
      type: 'REMOVE_EXPENSE',
      id: id
  });

export const startRemoveExpense = ({id} = {}) =>  {
    return (dispatch) => {
        const refToRemove = database.ref(`expenses/${id}`);
                return refToRemove
        .remove()
        .then( ()=> {
          dispatch(removeExpense({id}));
        });
    };
};   

//EDIT EXPENSE
export const editExpense= (id,updates) => ({
      type: "EDIT_EXPENSE",
      id: id,
      updates: updates
  });


export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

//export const startSetExpenses;
export const startSetExpenses = () => {
    return (dispatch) => {
        
        return database.ref('expenses')
          .once('value')
          .then((snapshot) => {
            const expenses = [];
            snapshot.forEach((childSnap) => {
              expenses.push({
                id: childSnap.key,
                ...childSnap.val()
                });
            })
            dispatch(setExpenses(expenses));
          })
    };
}


