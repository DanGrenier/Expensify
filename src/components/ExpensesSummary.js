import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';
import pluralize from 'pluralize';
import numeral from 'numeral';

 export const ExpensesSummary = ({expenseCount, expenseTotal}) => {
    return (
    <div>
      <h1>Viewing {pluralize('expense',expenseCount,true)} totalling {numeral(expenseTotal / 100).format('$0,0.00')}</h1>
    </div>
    );
 };


 const mapStateToProps = (state) => {
     const visibleExpenses = selectExpenses(state.expenses, state.filters);
     console.log(visibleExpenses);

     return {
         expenseCount: visibleExpenses.length,
         expenseTotal: selectExpensesTotal(visibleExpenses)
     }
 }


export default connect(mapStateToProps)(ExpensesSummary);


