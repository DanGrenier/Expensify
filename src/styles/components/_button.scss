.button {
    background: $light-blue;
    padding: $s-size;
    color: #fff;
    display: inline-block;
    font-family: Helvetica, Arial, sans-serif;
    font-size: $font-size-large;
    font-weight: 300;
    border: none;
    border-radius: 3px;
    text-decoration: none;
    line-height: 1;

}

.button--link {
    background: none;
}

.button--secondary{
    background: #888888;
}





























