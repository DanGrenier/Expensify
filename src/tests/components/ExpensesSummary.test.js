import React from 'react';
import { shallow } from 'enzyme';
import {ExpensesSummary} from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';


test('It should render ExpensesSummary for one expense', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={1} expenseTotal={12000} />);
    expect(wrapper).toMatchSnapshot();
})

test('It should render ExpensesSummary for multiple expenses', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={expenses.length} expenseTotal={120000} />);
    expect(wrapper).toMatchSnapshot();
})