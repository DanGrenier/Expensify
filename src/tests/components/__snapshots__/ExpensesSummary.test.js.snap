// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It should render ExpensesSummary for multiple expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        3 expenses
      </span>
       totalling 
      <span>
        $1,200.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`It should render ExpensesSummary for one expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1 expense
      </span>
       totalling 
      <span>
        $120.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
